dash-pipeline/dpapp/dash/dash.c

```diff
@@ -0,0 +1,212 @@
+                                unformat_input_t * input,
+                                vlib_cli_command_t * cmd)
+{
+  dash_main_t * sm = &dash_main;
+  u32 sw_if_index = ~0;
+  int enable_disable = 1;
+  int rv;
+
+  while (unformat_check_input (input) != UNFORMAT_END_OF_INPUT) {
+    if (unformat (input, "disable"))
+      enable_disable = 0;
+    else if (unformat (input, "%U", unformat_vnet_sw_interface,
+                       sm->vnet_main, &sw_if_index))
+      ;
+    else
+      break;
+  }
+
+  if (sw_if_index == ~0)
+    return clib_error_return (0, "Please specify an interface...");
+
+  rv = dash_enable_disable (sm, sw_if_index, enable_disable);
+
+  switch(rv) {
+  case 0:
+    break;
+
+  case VNET_API_ERROR_INVALID_SW_IF_INDEX:
+    return clib_error_return
+      (0, "Invalid interface, only works on physical ports");
+    break;
+
+  case VNET_API_ERROR_UNIMPLEMENTED:
+    return clib_error_return (0, "Device driver doesn't support redirection");
+    break;
+
+  default:
+    return clib_error_return (0, "dash_enable_disable returned %d",
+                              rv);
+  }
+  return 0;
+}
+
+/**
+ * @brief CLI command to enable/disable the dash plugin.
+ */
+VLIB_CLI_COMMAND (dash_set_command, static) = {
+    .path = "set dash",
+    .short_help =
+    "set dash <interface-name> [disable]",
+    .function = dash_cmd_set_enable_disable_fn,
+};
+
+/**
+ * @brief Plugin API message handler.
+ */
+static void vl_api_dash_enable_disable_t_handler
+(vl_api_dash_enable_disable_t * mp)
+{
+  vl_api_dash_enable_disable_reply_t * rmp;
+  dash_main_t * sm = &dash_main;
+  int rv;
+
+  rv = dash_enable_disable (sm, ntohl(mp->sw_if_index),
+                            (int) (mp->enable_disable));
+
+  REPLY_MACRO(VL_API_DASH_ENABLE_DISABLE_REPLY);
+}
+
+/* API definitions */
+#include <dash/dash.api.c>
+
+/**
+ * @brief Initialize the dash plugin.
+ */
+static clib_error_t * dash_init (vlib_main_t * vm)
+{
+  dash_main_t * sm = &dash_main;
+
+  sm->vnet_main =  vnet_get_main ();
+
+  /* Add our API messages to the global name_crc hash table */
+  sm->msg_id_base = setup_message_id_table ();
+
+  /* Reuse SECURE_DATA (0x876D) for dash metadata */
+  ethernet_register_input_type (vm, ETHERNET_TYPE_SECURE_DATA, dash_node.index);
+
+  dash_flow_table_init(dash_flow_table_get());
+
+  dash_sai_init();
+
+  return 0;
+}
+
+VLIB_INIT_FUNCTION (dash_init);
+
+/**
+ * @brief Hook the dash plugin into the VPP graph hierarchy.
+ */
+VNET_FEATURE_ARC_INIT (dash_pipeline, static) =
+{
+  .arc_name = "dash-pipeline",
+  .start_nodes = VNET_FEATURES ("dash-pipeline-input"),
+  .last_in_arc = "error-drop",
+  .arc_index_ptr = &dash_main.feature_arc_index,
+};
+
+static uword
+dash_timer_process (vlib_main_t * vm, vlib_node_runtime_t * rt, vlib_frame_t * f)
+{
+
+  int i;
+  f64 sleep_duration = 1.0;
+
+  unix_sleep (5.0); /* FIXME: delay 5s */
+
+  while (1)
+  {
+      /* FIXME: Use time-wheel per-worker thread later */
```

Comment:
Add comment here to highlight the threading model, because dash_flow_table doesn't have a lock.

Reply:
remove the for loop and fix me.

Reply:
fixed